
<!DOCTYPE html>
<html lang='en'>
<head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"/>
    <title>Tree - rpms/grub2 - src.fedoraproject.org</title>
        <link rel="shortcut icon" type="image/vnd.microsoft.icon"
        href="/theme/static/favicon.ico?version=5.10.0"/>
    <link href="/theme/static/fedora-bootstrap-1.3.0/fedora-bootstrap.min.css?version=5.10.0"
        type="text/css" rel="stylesheet" />
    <link href="/theme/static/fonts/fonts.css?version=5.10.0"
        rel="stylesheet" type="text/css" />
    <link href="/theme/static/fonts/hack_fonts/css/hack-extended.min.css?version=5.10.0"
        type="text/css" rel="stylesheet" />
    <link href="/theme/static/theme.css?version=5.10.0"
        type="text/css" rel="stylesheet" />

    <link type="text/css" rel="stylesheet" nonce="l1rmz83aeKYUfoZGtUY2SWH5Q" href="/static/vendor/font-awesome/font-awesome.css?version=5.10.0"/>
    <link type="text/css" rel="stylesheet" nonce="l1rmz83aeKYUfoZGtUY2SWH5Q" href="/static/pagure.css?version=5.10.0"/>
<link nonce="l1rmz83aeKYUfoZGtUY2SWH5Q" rel="stylesheet" href="/static/vendor/highlight.js/styles/github.css?version=5.10.0"/>
<link nonce="l1rmz83aeKYUfoZGtUY2SWH5Q" rel="stylesheet" href="/static/vendor/highlightjs-line-numbers/highlightjs-line-numbers.min.css?version=5.10.0"/>
<style nonce="l1rmz83aeKYUfoZGtUY2SWH5Q">
  .hljs {
    background: #fff;
  }
</style>
  </head>
  <body id="home">
    
    <!-- start masthead -->
    <nav class="navbar navbar-light masthead p-0 navbar-expand">
      <div class="container">
        <a href="/" class="navbar-brand">
        <img height=40px src="/theme/static/pagure-logo.png?version=5.10.0"
             alt="pagure Logo" id="pagureLogo"/>
        </a>
        <ul class="navbar-nav ml-auto">
          <li class="nav-item">
            <a class="btn btn-primary" href="/login/?next=https://src.fedoraproject.org/rpms/grub2/blob/master/f/0131-Add-grub-set-bootflag-utility.patch">Log In</a>
          </li>
        </ul>
      </div>
    </nav>
    <!-- close masthead-->

    <div class="bodycontent">


<div class="bg-light border border-bottom pt-3">
  <div class="container">
    <div class="row mb-3">
      <div class="col-6">
        <div class="row">
        <div class="col-auto pr-0">
            <h3>
<i class="fa fa-archive text-muted"></i></h3>
        </div>
        <div class="col-auto pl-2">
            <h3 class="mb-0">
<a href="/projects/rpms/%2A">rpms</a>&nbsp;/&nbsp;<a href="/rpms/grub2"><strong>grub2</strong></a>
            </h3>
        </div>
        </div>
      </div>
      <div class="col-6 text-right">
        <div class="btn-group">
        <div class="btn-group">
        <a href="#"
            class="btn btn-sm dropdown-toggle btn-outline-primary"
            data-toggle="dropdown" id="watch-button">
          <i class="fa fa-clone fa-fw"></i>
          <span>Clone</span>
        </a>
        <div class="dropdown-menu dropdown-menu-right">
          <div class="m-3" id="source-dropdown" class="pointer">
            <div>
              <h5><strong>Source Code</strong></h5>

              <div class="form-group">
                <div class="input-group input-group-sm">
                  <div class="input-group-prepend"><span class="input-group-text">GIT</span></div>
                  <input class="form-control bg-white select-on-focus" type="text" value="https://src.fedoraproject.org/rpms/grub2.git" readonly>
                </div>
              </div>
            </div>
          </div>
        </div>

      </div>
    </div>
  </div>
</div>

<ul class="nav nav-tabs nav-small border-bottom-0">
  <li class="nav-item mr-2 text-dark">
    <a class="nav-link active" href="/rpms/grub2">
        <i class="fa fa-code fa-fw text-muted"></i>
        <span class="d-none d-md-inline">Source</span>
    </a>
  </li>

    <li class="nav-item mr-2 text-dark">
        <a class="nav-link" href="https://bugzilla.redhat.com/buglist.cgi?bug_status=NEW&bug_status=ASSIGNED&classification=Fedora&product=Fedora&product=Fedora EPEL&component=grub2">
          <i class="fa fa-fw text-muted fa-exclamation-circle"></i>
          <span class="d-none d-md-inline">Issues&nbsp;</span>
          <span class="fa fa-external-link"></span>
        </a>
    </li>



    <li class="nav-item mr-2 text-dark">
      <a class="nav-link" href="/rpms/grub2/stats">
          <i class="fa fa-line-chart fa-fw text-muted"></i>
          <span class="d-none d-md-inline">Stats</span>
      </a>
    </li>


</ul>
  </div>
</div>

<div class="container pt-5 repo-body-container">
  <div class="row">
    <div class="col-2">
<nav class="nav nav-tabs nav-sidetabs flex-column">
  <a class=
      "nav-link nowrap
"
      href="/rpms/grub2">
      <i class="fa fa-home text-muted fa-fw"></i>&nbsp;<span class="d-none d-md-inline">Overview</span>
  </a>
  <a class=
    "nav-link nowrap
 active"
    href="/rpms/grub2/tree/master">
    <i class="fa fa-file-code-o text-muted fa-fw"></i>&nbsp;Files
  </a>
  <a class=
    "nav-link nowrap
"
    href="/rpms/grub2/commits/master">
    <i class="fa fa-list-alt text-muted fa-fw" data-glyph="spreadsheet"></i>&nbsp;Commits
  </a>
  <a class=
    "nav-link nowrap
"
    href="/rpms/grub2/branches?branchname=master">
    <i class="fa fa-random text-muted fa-fw"></i>&nbsp;Branches
  </a>
  <a class=
    "nav-link nowrap
"
    href="/rpms/grub2/forks">
    <i class="fa fa-code-fork text-muted fa-fw"></i>&nbsp;Forks
  </a>
  <a class=
    "nav-link nowrap
"
    href="/rpms/grub2/releases">
    <i class="fa fa-tags text-muted fa-fw"></i>&nbsp;Releases
  </a>

  <div class="col-xs-2 line-height-1"></div>
  <h6>Monitoring status:</h6>
  <div class="btn-group">
    <button title="Monitoring status" class="btn btn-sm btn-outline-primary disabled"
        id="monitoring-button">
      <i id="monitoring-icon" class="fa fa-fw fa-eye"></i>
      <span id="monitoring-label" class="fa fa-circle-o-notch fa-spin fa-1x fa-fw"></span>
    </button>
  </div>


  <div class="pt-3">
    <div class="col-xs-2 line-height-1">
    <h6>Bugzilla Assignee:</h6>
      <dl>
        <dt>Fedora: </dt>
        <dd id="fedora_assignee_txt">
        pjones
        </dd>
        <dt>EPEL: </dt>
        <dd id="epel_assignee_txt">
        pjones
        </dd>
      </dl>
    </div>
  </div>

  <div class="modal fade" id="modal_assignee" tabindex="-1"
          role="dialog" aria-labelledby="Bugzilla assignee" aria-hidden="true">
    <div class="modal-dialog" role="document">
      <div class="modal-content">
        <div class="modal-header">
          <h4 class="modal-title">Bugzilla Assignee</h4>
          <button type="button" class="close" data-dismiss="modal" aria-label="Close">
            <span aria-hidden="true">&times;</span>
            <span class="sr-only">Close</span>
          </button>
        </div>
        <form id="change_assignees">
          <div class="modal-body">
            <label for="fedora_assignee">Fedora</label>
            <input title="Default assignee for Fedora in bugzilla - Empty input resets to default"
              class="form-control" name="fedora_assignee" id="fedora_assignee" value="pjones"/>
            <label for="epel_assignee">EPEL</label>
            <input title="Default assignee for EPEL in bugzilla (if applicable) - Empty input resets to default"
              class="form-control" name="epel_assignee" id="epel_assignee" value="pjones" />
            <p class="pt-2">
              These two fields allow to specify a different default assignee for ticket opened against
              this package in bugzilla. Note: The EPEL field is
              always displayed for packages in the 'rpms' namespace regardless of whether it
              is used in bugzilla or not.            </p>
          </div>
          <div class="modal-footer">
            <button class="btn btn-secondary" type="button" title="Update bugzilla overrides" id="reset_assignees">
               Reset to defaults
            </button>
            <button class="btn btn-primary" type="submit" title="Update bugzilla overrides" id="update_assignees">
               Update
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>

  <script type="text/javascript" nonce="l1rmz83aeKYUfoZGtUY2SWH5Q">
    window.addEventListener('load', function() {
      set_up_monitoring = function(status){
        var _label = "Disabled"
        if (status === "monitoring") {
          _label = "Monitoring";
          $("#monitoring-icon").attr("class", "fa fa-fw fa-eye")
        } else if (status === "monitoring-with-scratch") {
          _label = "Scratch builds"
          $("#monitoring-icon").attr("class", "fa fa-fw fa-eye")
        } else {
          $("#monitoring-icon").attr("class", "fa fa-fw fa-eye-slash")
        }

        $("#monitoring-label").text(_label);
        $("#monitoring-label").removeClass("fa fa-circle-o-notch fa-spin fa-1x fa-fw");
      }

      $.ajax({
        url: "/_dg/anitya/rpms/grub2",
        type: 'GET',
        dataType: 'json',
        success: function(res) {
          console.log(res);
          set_up_monitoring(res.monitoring)
        }
      });


      $("#reset_assignees").on('click', function(){
        $('#fedora_assignee').val('');
        $('#epel_assignee').val('');
        $("#change_assignees").submit();
        return false;
      });

      $("#change_assignees").on('submit',  function(){
        $('html').css('cursor', 'progress');
        $('#reset_assignees').attr('disabled', true);
        $('#update_assignees').attr('disabled', true);
        $('#update_assignees').text('Updating...');
        $.ajax({
          url: "/_dg/bzoverrides/rpms/grub2",
          type: 'POST',
          dataType: 'json',
          data: {
             'epel_assignee': $('#epel_assignee').val(),
             'fedora_assignee': $('#fedora_assignee').val()
          },
          success: function(res) {
            $("#fedora_assignee_txt").text(res.fedora_assignee);
            $("#epel_assignee_txt").text(res.epel_assignee);
            $('#modal_assignee').modal('hide');
            $('#reset_assignees').attr('disabled', false);
            $('#update_assignees').attr('disabled', false);
            $('#update_assignees').text('Update');
            $('html').css('cursor', 'default');
            console.log("Successfully changed the bugzilla assignees");
            return false;
          },
          error: function(res) {
            var msg = '';
            if(res.responseJSON.errors){
              msg = ': ' + res.responseJSON.errors.join(', ');
            }
            alert("Unable to update the bugzilla assignee(s)" + msg);
            $('html').css('cursor', 'default');
            $('#reset_assignees').attr('disabled', false);
            $('#update_assignees').attr('disabled', false);
            $('#update_assignees').text('Update');
            return false;
          }
        })
        return false;
      });


    });
  </script>

</nav>    </div>
    <div class="col-10">
  <div class="row mb-1">
    <div class="col-sm-6">
    <h3>
      Files
    </h3>
    </div>

    <div class="col-sm-6">
      <div class="float-right">
          <div class="btn-group">
            <a href="#" class="btn btn-outline-light border-secondary text-dark btn-sm dropdown-toggle"
                    data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
                    <span class="fa fa-random fa-fw"></span> Branch: <span class="font-weight-bold">master</span>
          </a>
            <div class="dropdown-menu dropdown-menu-right">
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/el5">
                      <span class="fa fa-random fa-fw"></span> <span class="">el5</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/el6">
                      <span class="fa fa-random fa-fw"></span> <span class="">el6</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/f10">
                      <span class="fa fa-random fa-fw"></span> <span class="">f10</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/f11">
                      <span class="fa fa-random fa-fw"></span> <span class="">f11</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/f12">
                      <span class="fa fa-random fa-fw"></span> <span class="">f12</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/f13">
                      <span class="fa fa-random fa-fw"></span> <span class="">f13</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/f14">
                      <span class="fa fa-random fa-fw"></span> <span class="">f14</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/f15">
                      <span class="fa fa-random fa-fw"></span> <span class="">f15</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/f16">
                      <span class="fa fa-random fa-fw"></span> <span class="">f16</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/f17">
                      <span class="fa fa-random fa-fw"></span> <span class="">f17</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/f18">
                      <span class="fa fa-random fa-fw"></span> <span class="">f18</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/f19">
                      <span class="fa fa-random fa-fw"></span> <span class="">f19</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/f20">
                      <span class="fa fa-random fa-fw"></span> <span class="">f20</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/f21">
                      <span class="fa fa-random fa-fw"></span> <span class="">f21</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/f22">
                      <span class="fa fa-random fa-fw"></span> <span class="">f22</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/f23">
                      <span class="fa fa-random fa-fw"></span> <span class="">f23</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/f24">
                      <span class="fa fa-random fa-fw"></span> <span class="">f24</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/f25">
                      <span class="fa fa-random fa-fw"></span> <span class="">f25</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/f25-bls">
                      <span class="fa fa-random fa-fw"></span> <span class="">f25-bls</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/f25-bls3">
                      <span class="fa fa-random fa-fw"></span> <span class="">f25-bls3</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/f26">
                      <span class="fa fa-random fa-fw"></span> <span class="">f26</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/f27">
                      <span class="fa fa-random fa-fw"></span> <span class="">f27</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/f28">
                      <span class="fa fa-random fa-fw"></span> <span class="">f28</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/f29">
                      <span class="fa fa-random fa-fw"></span> <span class="">f29</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/f30">
                      <span class="fa fa-random fa-fw"></span> <span class="">f30</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/f31">
                      <span class="fa fa-random fa-fw"></span> <span class="">f31</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/f32">
                      <span class="fa fa-random fa-fw"></span> <span class="">f32</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/f9">
                      <span class="fa fa-random fa-fw"></span> <span class="">f9</span>
                  </a>
                  <a class="dropdown-item pl-1 active" href="/rpms/grub2/tree/master">
                      <span class="fa fa-random fa-fw"></span> <span class="font-weight-bold">master</span>
                  </a>
                  <a class="dropdown-item pl-1 " href="/rpms/grub2/tree/splitarch4">
                      <span class="fa fa-random fa-fw"></span> <span class="">splitarch4</span>
                  </a>
            </div>
          </div>
    </div>
  </div>
  </div>
    <div class="card mb-3">
      <div class="card-header">
        <ol class="breadcrumb p-0 bg-transparent mb-0">
          <li class="breadcrumb-item">
            <a href="/rpms/grub2/tree/master">
              <span class="fa fa-random">
              </span>&nbsp; master
            </a>
          </li>
          <li class="active breadcrumb-item">
            <span class="fa fa-file" data-glyph="">
            </span>&nbsp; 0131-Add-grub-set-bootflag-utility.patch
          </li>
        </ol>
      </div>

  <div class="card-body p-0">
            <div class="bg-light border text-right pr-2">
                <form class="btn btn-sm" method="POST" name="fork_project"
                    action="/fork_edit/rpms/grub2/edit/master/f/0131-Add-grub-set-bootflag-utility.patch">
                    <button class="btn btn-sm btn-secondary fork_project_btn">
                            Fork and Edit
                    </button>
                    
                </form>

                <a class="btn btn-secondary btn-sm" href="/rpms/grub2/blob/master/f/0131-Add-grub-set-bootflag-utility.patch" title="View as blob">Blob</a>

                <a class="btn btn-secondary btn-sm" href="/rpms/grub2/blame/0131-Add-grub-set-bootflag-utility.patch?identifier=master" title="View git blame">Blame</a>

                <a class="btn btn-secondary btn-sm" href="/rpms/grub2/history/0131-Add-grub-set-bootflag-utility.patch?identifier=master" title="View git log for this file">History</a>

                <a class="btn btn-secondary btn-sm" href="/rpms/grub2/raw/master/f/0131-Add-grub-set-bootflag-utility.patch" title="View as raw">Raw</a>
            </div>

        <pre class="syntaxhighlightblock"><code class="lang-diff">From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Hans de Goede &lt;hdegoede@redhat.com&gt;
Date: Tue, 12 Jun 2018 13:25:16 +0200
Subject: [PATCH] Add grub-set-bootflag utility

This commit adds a new grub-set-bootflag utility, which can be used
to set known bootflags in the grubenv: boot_success or menu_show_once.

grub-set-bootflag is different from grub-editenv in 2 ways:

1) It is intended to be executed by regular users so must be installed
as suid root. As such it is written to not use any existing grubenv
related code for easy auditing.

It can&#39;t be executed through pkexec because we want to call it under gdm
and pkexec does not work under gdm due the gdm user having /sbin/nologin
as shell.

2) Since it can be executed by regular users it only allows setting
(assigning a value of 1 to) bootflags which it knows about. Currently
those are just boot_success and menu_show_once.

This commit also adds a couple of example systemd and files which show
how this can be used to set boot_success from a user-session:

docs/grub-boot-success.service
docs/grub-boot-success.timer

The 2 grub-boot-success.systemd files should be placed in /lib/systemd/user
and a symlink to grub-boot-success.timer should be added to
/lib/systemd/user/timers.target.wants.

Signed-off-by: Hans de Goede &lt;hdegoede@redhat.com&gt;
---
 Makefile.util.def              |   7 ++
 util/grub-set-bootflag.c       | 160 +++++++++++++++++++++++++++++++++++++++++
 conf/Makefile.extra-dist       |   3 +
 docs/grub-boot-success.service |   6 ++
 docs/grub-boot-success.timer   |   6 ++
 util/grub-set-bootflag.1       |  20 ++++++
 6 files changed, 202 insertions(+)
 create mode 100644 util/grub-set-bootflag.c
 create mode 100644 docs/grub-boot-success.service
 create mode 100644 docs/grub-boot-success.timer
 create mode 100644 util/grub-set-bootflag.1

diff --git a/Makefile.util.def b/Makefile.util.def
index 89a9da1b9f7..125ad6209b2 100644
--- a/Makefile.util.def
+++ b/Makefile.util.def
@@ -1451,3 +1451,10 @@ program = {
   ldadd = grub-core/lib/gnulib/libgnu.a;
   ldadd = &#39;$(LIBINTL) $(LIBDEVMAPPER) $(LIBZFS) $(LIBNVPAIR) $(LIBGEOM)&#39;;
 };
+
+program = {
+  name = grub-set-bootflag;
+  installdir = sbin;
+  mansection = 1;
+  common = util/grub-set-bootflag.c;
+};
diff --git a/util/grub-set-bootflag.c b/util/grub-set-bootflag.c
new file mode 100644
index 00000000000..bb198f02351
--- /dev/null
+++ b/util/grub-set-bootflag.c
@@ -0,0 +1,160 @@
+/* grub-set-bootflag.c - tool to set boot-flags in the grubenv. */
+/*
+ *  GRUB  --  GRand Unified Bootloader
+ *  Copyright (C) 2018 Free Software Foundation, Inc.
+ *
+ *  GRUB is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  GRUB is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with GRUB.  If not, see &lt;http://www.gnu.org/licenses/&gt;.
+ */
+
+/*
+ * NOTE this gets run by users as root (through pkexec), so this does not
+ * use any grub library / util functions to allow for easy auditing.
+ * The grub headers are only included to get certain defines.
+ */
+
+#include &lt;config-util.h&gt;     /* For *_DIR_NAME defines */
+#include &lt;grub/types.h&gt;
+#include &lt;grub/lib/envblk.h&gt; /* For GRUB_ENVBLK_DEFCFG define */
+#include &lt;errno.h&gt;
+#include &lt;stdio.h&gt;
+#include &lt;string.h&gt;
+#include &lt;unistd.h&gt;
+
+#define GRUBENV &#34;/&#34; GRUB_BOOT_DIR_NAME &#34;/&#34; GRUB_DIR_NAME &#34;/&#34; GRUB_ENVBLK_DEFCFG
+#define GRUBENV_SIZE 1024
+
+const char *bootflags[] = {
+  &#34;boot_success&#34;,
+  &#34;menu_show_once&#34;,
+  NULL
+};
+
+static void usage(void)
+{
+  int i;
+
+  fprintf (stderr, &#34;Usage: &#39;grub-set-bootflag &lt;bootflag&gt;&#39;, where &lt;bootflag&gt; is one of:\n&#34;);
+  for (i = 0; bootflags[i]; i++)
+    fprintf (stderr, &#34;  %s\n&#34;, bootflags[i]);
+}
+
+int main(int argc, char *argv[])
+{
+  /* NOTE buf must be at least the longest bootflag length + 4 bytes */
+  char env[GRUBENV_SIZE + 1], buf[64], *s;
+  const char *bootflag;
+  int i, len, ret;
+  FILE *f;
+
+  if (argc != 2)
+    {
+      usage();
+      return 1;
+    }
+
+  for (i = 0; bootflags[i]; i++)
+    if (!strcmp (argv[1], bootflags[i]))
+      break;
+  if (!bootflags[i])
+    {
+      fprintf (stderr, &#34;Invalid bootflag: &#39;%s&#39;\n&#34;, argv[1]);
+      usage();
+      return 1;
+    }
+
+  bootflag = bootflags[i];
+  len = strlen (bootflag);
+
+  f = fopen (GRUBENV, &#34;r&#34;);
+  if (!f)
+    {
+      perror (&#34;Error opening &#34; GRUBENV &#34; for reading&#34;);
+      return 1;     
+    }
+
+  ret = fread (env, 1, GRUBENV_SIZE, f);
+  fclose (f);
+  if (ret != GRUBENV_SIZE)
+    {
+      errno = EINVAL;
+      perror (&#34;Error reading from &#34; GRUBENV);
+      return 1;     
+    }
+
+  /* 0 terminate env */
+  env[GRUBENV_SIZE] = 0;
+
+  if (strncmp (env, GRUB_ENVBLK_SIGNATURE, strlen (GRUB_ENVBLK_SIGNATURE)))
+    {
+      fprintf (stderr, &#34;Error invalid environment block\n&#34;);
+      return 1;
+    }
+
+  /* Find a pre-existing definition of the bootflag */
+  s = strstr (env, bootflag);
+  while (s &amp;&amp; s[len] != &#39;=&#39;)
+    s = strstr (s + len, bootflag);
+
+  if (s &amp;&amp; ((s[len + 1] != &#39;0&#39; &amp;&amp; s[len + 1] != &#39;1&#39;) || s[len + 2] != &#39;\n&#39;))
+    {
+      fprintf (stderr, &#34;Pre-existing bootflag &#39;%s&#39; has unexpected value\n&#34;, bootflag);
+      return 1;     
+    }
+
+  /* No pre-existing bootflag? -&gt; find free space */
+  if (!s)
+    {
+      for (i = 0; i &lt; (len + 3); i++)
+        buf[i] = &#39;#&#39;;
+      buf[i] = 0;
+      s = strstr (env, buf);
+    }
+
+  if (!s)
+    {
+      fprintf (stderr, &#34;No space in grubenv to store bootflag &#39;%s&#39;\n&#34;, bootflag);
+      return 1;     
+    }
+
+  /* The grubenv is not 0 terminated, so memcpy the name + &#39;=&#39; , &#39;1&#39;, &#39;\n&#39; */
+  snprintf(buf, sizeof(buf), &#34;%s=1\n&#34;, bootflag);
+  memcpy(s, buf, len + 3);
+
+  /* &#34;r+&#34;, don&#39;t truncate so that the diskspace stays reserved */
+  f = fopen (GRUBENV, &#34;r+&#34;);
+  if (!f)
+    {
+      perror (&#34;Error opening &#34; GRUBENV &#34; for writing&#34;);
+      return 1;     
+    }
+
+  ret = fwrite (env, 1, GRUBENV_SIZE, f);
+  if (ret != GRUBENV_SIZE)
+    {
+      perror (&#34;Error writing to &#34; GRUBENV);
+      return 1;     
+    }
+
+  ret = fflush (f);
+  if (ret)
+    {
+      perror (&#34;Error flushing &#34; GRUBENV);
+      return 1;     
+    }
+
+  fsync (fileno (f));
+  fclose (f);
+
+  return 0;
+}
diff --git a/conf/Makefile.extra-dist b/conf/Makefile.extra-dist
index 58d7d9540be..375b1bf9b02 100644
--- a/conf/Makefile.extra-dist
+++ b/conf/Makefile.extra-dist
@@ -14,6 +14,9 @@ EXTRA_DIST += util/import_unicode.py
 EXTRA_DIST += docs/autoiso.cfg
 EXTRA_DIST += docs/grub.cfg
 EXTRA_DIST += docs/osdetect.cfg
+EXTRA_DIST += docs/org.gnu.grub.policy
+EXTRA_DIST += docs/grub-boot-success.service
+EXTRA_DIST += docs/grub-boot-success.timer
 
 EXTRA_DIST += conf/i386-cygwin-img-ld.sc
 
diff --git a/docs/grub-boot-success.service b/docs/grub-boot-success.service
new file mode 100644
index 00000000000..80e79584c91
--- /dev/null
+++ b/docs/grub-boot-success.service
@@ -0,0 +1,6 @@
+[Unit]
+Description=Mark boot as successful
+
+[Service]
+Type=oneshot
+ExecStart=/usr/sbin/grub2-set-bootflag boot_success
diff --git a/docs/grub-boot-success.timer b/docs/grub-boot-success.timer
new file mode 100644
index 00000000000..5d8fcba21aa
--- /dev/null
+++ b/docs/grub-boot-success.timer
@@ -0,0 +1,6 @@
+[Unit]
+Description=Mark boot as successful after the user session has run 2 minutes
+ConditionUser=!@system
+
+[Timer]
+OnActiveSec=2min
diff --git a/util/grub-set-bootflag.1 b/util/grub-set-bootflag.1
new file mode 100644
index 00000000000..57801da22a0
--- /dev/null
+++ b/util/grub-set-bootflag.1
@@ -0,0 +1,20 @@
+.TH GRUB-SET-BOOTFLAG 1 &#34;Tue Jun 12 2018&#34;
+.SH NAME
+\fBgrub-set-bootflag\fR \(em Set a bootflag in the GRUB environment block.
+
+.SH SYNOPSIS
+\fBgrub-set-bootflag\fR &lt;\fIBOOTFLAG\fR&gt;
+
+.SH DESCRIPTION
+\fBgrub-set-bootflag\fR is a command line to set bootflags in GRUB&#39;s
+stored environment.
+
+.SH COMMANDS
+.TP
+\fBBOOTFLAG\fR
+.RS 7
+Bootflag to set, one of \fIboot_success\fR or \fIshow_menu_once\fR.
+.RE
+
+.SH SEE ALSO
+.BR &#34;info grub&#34;
</code></pre>
  </div>
 </div> <!-- end .card-->

</div>
</div>
</div>
    </div>

        <div class="footer pt-4 text-white">
        <div class="container">
            <div class="d-flex align-items-center">
                <div>
                    <div>Powered by <a href="https://pagure.io/pagure" class="notblue">Pagure</a> 5.10.0</div>
                    <div>
                        <a href="https://docs.pagure.org/pagure/usage/index.html" class="notblue">Documentation</a> &bull;
                        <a href="https://pagure.io/pagure/new_issue" class="notblue">File an Issue</a> &bull;
                        <a href="/about">About this Instance</a> &bull;
                        <a href="/ssh_info" class="notblue">SSH Hostkey/Fingerprint</a>
                    </div>
                </div>
                <div class="ml-auto text-right">
                    <div>&copy; Red Hat, Inc. and others.</div>
                </div>
            </div>
        </div>
    </div>


    <script type="text/javascript" nonce="l1rmz83aeKYUfoZGtUY2SWH5Q" src="/static/vendor/jquery/jquery.min.js?version=5.10.0"></script>

    <script src="/static/vendor/bootstrap/bootstrap.bundle.min.js?version=5.10.0"></script>

    <script type="text/javascript" nonce="l1rmz83aeKYUfoZGtUY2SWH5Q">
      $('[data-toggle="tooltip"]').tooltip({placement : 'bottom'});
      $(".cancel_btn").click(function() {
        history.back();
      });
    </script>

<script type="text/javascript"  nonce="l1rmz83aeKYUfoZGtUY2SWH5Q" src="/static/vendor/lazyload/lazyload.min.js?version=5.10.0"></script>

<script type="text/javascript" nonce="l1rmz83aeKYUfoZGtUY2SWH5Q">
window.addEventListener("load", function(event) {
    lazyload();
});
</script>

<script type="text/javascript" nonce="l1rmz83aeKYUfoZGtUY2SWH5Q">
$("#giturl-toggle").on('click', function(event){
  event.stopPropagation();
  $("#giturl-more").toggle();
  $("#giturl-toggle").hide();
})

$(".fork_project_btn").click(function() {
  $('#fork_project').submit();
});

$(".select-on-focus").on("focus", function() {
  $(this).select();
});

</script>

<script type="text/javascript" nonce="l1rmz83aeKYUfoZGtUY2SWH5Q" src="/static/vendor/highlight.js/highlight.pack.js?version=5.10.0"></script>
<script type="text/javascript" nonce="l1rmz83aeKYUfoZGtUY2SWH5Q" src="/static/vendor/highlightjs-line-numbers/highlightjs-line-numbers.min.js?version=5.10.0"></script>
<script type="text/javascript" nonce="l1rmz83aeKYUfoZGtUY2SWH5Q" src="/static/vendor/highlight.js/spec.js?version=5.10.0"></script>

<script type="text/javascript" nonce="l1rmz83aeKYUfoZGtUY2SWH5Q">
  $(document).ready(function() {
  $('.fork_project_btn').click($("[name=fork_project]").submit);

  $('pre.syntaxhighlightblock code').each(function(i, block) {
    hljs.highlightBlock(block);
    hljs.lineNumbersBlock(block);
  });

  var cls = "highlighted-line";
  var lines = location.hash.substr(2).split('-').map(function (x) { return parseInt(x, 10) });
  if (! isNaN(lines[0]))
  {
    for (var i = lines[lines.length - 1]; i >= lines[0]; i--) {
      $('#_' + i).parent().parent().addClass(cls);
    }
    setTimeout(function(){
      $("#_" + lines[0]).get(0).scrollIntoView({behavior: "instant", block: "start", inline: "nearest"});
    }, 50);
  }
});
</script>

<script type="text/javascript" nonce="l1rmz83aeKYUfoZGtUY2SWH5Q">

  function updateHighlight() {
    var cls = "highlighted-line";
    $('.' + cls).removeClass(cls)
    if (location.hash !== '') {
      var lines = location.hash.substr(2).split('-').map(function (x) { return parseInt(x, 10) });
      for (var i = lines[lines.length - 1]; i >= lines[0]; i--) {
        $('[data-line-number=' + i + ']').closest('tr').addClass(cls);
      }
      return lines;
    }
    return [];
  }
  $(window).on('hashchange', updateHighlight);
  var selected = [];
  $("[data-line-number]").click(function (ev) {
    var line = $(this).attr('data-line-number');
    if (ev.shiftKey) {
      selected = selected.slice(-1).concat(line);
    } else {
      selected = [line];
    }

    var hash = '_' + selected[0];
    if (selected.length === 2) {
      hash = '_' + Math.min(selected[0], selected[1]) + '-' + Math.max(selected[0], selected[1]);
    }
    window.location.hash = hash;
    return false;
  });

</script>


</body>
</html>